src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1228 @@
+        // value so we push it back onto the heap before loading the next data.
+        // In the second case, it is possible that mTop is returned again (if it
+        // is still the best offer) or that a different offer is returned.
+        if (mTop->valid())
+        {
+            throw std::runtime_error("Previous best offer is still valid");
+        }
+        if (mTop->ignoreInvalid().entry())
+        {
+            auto const& offer = mTop->ignoreInvalid().entry()->data.offer();
+            if (offer.buying == mBuying && offer.selling == mSelling)
+            {
+                mInMemory.push(mTop);
+            }
+        }
+        mTop.reset();
+    }
+
+    while (!mTop)
+    {
+        loadFromDatabaseIfNecessary();
+        if (mInMemory.empty())
+        {
+            if (mFromDatabase.empty())
+            {
+                return nullptr;
+            }
+            else
+            {
+                // In this case, there are offers (with the correct assets) in
+                // memory. It is possible, then, that the best offer in the
+                // database is already loaded in memory. If that is the case, we
+                // must skip the result from the database. Then we repeat the
+                // entire process (not just this branch) to find the next data.
+                mTop = mFromDatabase.front();
+                mFromDatabase.erase(mFromDatabase.begin());
+                if (mLedgerState.isInMemory(LedgerEntryKey(*mTop->entry())))
+                {
+                    mTop.reset();
+                }
+            }
+        }
+        else if (mFromDatabase.empty() ||
+                 compareOffers(mInMemory.top(), mFromDatabase.front()))
+        {
+            mTop = mInMemory.top();
+            mInMemory.pop();
+        }
+        else
+        {
+            // In this case, there are offers (with the correct assets) in
+            // memory. It is possible, then, that the best offer in the
+            // database is already loaded in memory. If that is the case, we
+            // must skip the result from the database. Then we repeat the
+            // entire process (not just this branch) to find the next data.
+            mTop = mFromDatabase.front();
+            mFromDatabase.erase(mFromDatabase.begin());
+            if (mLedgerState.isInMemory(LedgerEntryKey(*mTop->entry())))
+            {
+                mTop.reset();
+            }
+        }
+    }
+
+    auto const& entry = mTop->ignoreInvalid().entry();
+    auto key = LedgerEntryKey(*entry);
+    auto iter = mLedgerState.mState.find(key);
+    if (iter != mLedgerState.mState.end())
+    {
+        auto const& previous = iter->second->ignoreInvalid().previousEntry();
+        mTop = mLedgerState.makeStateEntry(entry, previous);
+    }
+    else
+    {
+        mTop = mLedgerState.makeStateEntry(entry, entry);
+    }
+    mLedgerState.mState[key] = mTop;
```

Comment:
I would prefer not duplicating the code that lives in LegderState proper:
in this case, I *think* that you got `previous` wrong for example... (should be doing what `LedgerState::loadHelper` does : use whatever value `parent` says)

Easiest might be to just push values from `loadBestOffersFromDatabase` into the cache, and just call `LedgerState::loadHelper` (that will hit the cache).

Reply:
we really need more meta data tests...